text/nodejs/0014-yarn-berry.md

```diff
@@ -0,0 +1,246 @@
+# Support for Yarn "Berry" 
+
+## Summary
+
+The Paketo Node.js Buildpack supports building apps which use
+[Yarn](https://yarnpkg.com/) as a package/project management tool. Over the
+last few years, the developers of Yarn have released new major versions (v2 and
+greater), collectively called Yarn "Berry", which represent a completely new
+direction for managing dependencies in Node.js projects. Chief among the
+changes is the introduction of [Plug 'n
+Play](https://yarnpkg.com/features/pnp), a system which removes the reliance on
+`node_modules` and can eliminate the need to run `yarn install` at all. The
+Yarn implementation buildpacks do not currently support building apps which use
+Berry.  
+
+There have been quite a number of changes in the behaviour of Yarn with the
+advent of Berry. Along with performance improvements and optimizations, many
+CLI operations and methods of configuration have been deprecated or removed,
+including those crucial to the functioning of the buildpack as it exists today. 
+
+* _Throughout this document, Yarn Berry (> 2.x) and Yarn Classic (1.x) will be
+  referred to as "Berry" and "Classic" respectively._
+
+## Proposal
+
+Berry seems to represent the direction in which the project is headed, as
+[stated by its lead
+maintainer](https://dev.to/arcanis/introducing-yarn-2-4eh1#conclusion). Given
+this, and the extent of the changes to the CLI, this RFC proposes that Berry
+workflows be included in the `yarn-install` buildpack under two separate codepaths/packages.
+
+## Motivation
+
+As developers continue to adopt Berry, demand for the Node.js buildpack to
+support it has steadily increased.
+
+
+## Implementation
+
+
+### API
+
+`yarn-install`
+
+Provides: `node_pkgs`
+Requires: `node`, `yarn` during `build`
+
+The Node.js buildpacks use `node_modules` as a keyword for providing/requiring
+packages in the build plan across the board. Since Berry explicitly excludes
+`node_modules`, however, it seems reasonable to instead use a generic term for
+node packages such as `node_pkgs`.
+
+The buildpacks affected by this change to the `node_modules` build plan dependency are:
+
+- `npm-install`
+- `yarn-install`
+- `npm-start`
+- `yarn-start`
+- `node-start`
+- `node-run-script`
+- `node-module-bom`
+- `rails-assets`
```

Comment:
Can you provide some more of the rational behind this choice? I'm not seeing how we don't end up in a state where its either confusing to Berry users or confusing to non-Berry users about what is being provided here.

Reply:
To me, it'd be more intuitive if the yarn classic and yarn berry paths resulted in two slightly different build plans, since they do involve dealing with different sets of artifacts. 

I'd be in favor of the yarn-install buildpack providing `node_modules` if it plans on running `yarn install` and `yarn_cache` if it doesn't. Then the start command buildpacks could require one `Or` the other.

Reply:
@ryanmoran does @fg-j's comment more closely align with what you'd like to see? I'm fine with additional logic for the sake of clarity.